File: backend/schemas/market_panier.py
from typing import Any, Optional

from pydantic import BaseModel, Field


class PanierItemIn(BaseModel):
    ingredient_nom: str
    quantite: float = Field(gt=0)
    unite: str = "g"
    ingredient_id: Optional[str] = None


class PanierCheckRequest(BaseModel):
    items: list[PanierItemIn] = Field(min_length=1)
    budget: float = Field(gt=0)
    quartier: Optional[str] = None
    lat: Optional[float] = None
    lon: Optional[float] = None


class PanierSwapSuggestion(BaseModel):
    ingredient_nom: str
    alternative: str
    economie_estimee: float
    raison: str


class PanierCheckResponse(BaseModel):
    cout_estime: float
    budget: float
    ecart: float
    statut: str  # sous_budget | au_budget | over_budget
    swaps: list[PanierSwapSuggestion] = Field(default_factory=list)


class OneTripRequest(BaseModel):
    items: list[PanierItemIn] = Field(min_length=1)
    lat: float
    lon: float
    rayon_km: float = Field(default=15, gt=0, le=50)
    budget: Optional[float] = Field(default=None, gt=0)
    profil_id: Optional[str] = None


class OneTripStopItem(BaseModel):
    ingredient_id: str
    ingredient_nom: str
    quantite: float
    unite: str
    prix_unitaire: float
    cout_estime: float


class OneTripPdv(BaseModel):
    id: str
    nom: str
    type: str
    latitude: float
    longitude: float


class OneTripStop(BaseModel):
    point_de_vente: OneTripPdv
    distance_km: float
    cout_estime: float
    items: list[OneTripStopItem]


class OneTripManquant(BaseModel):
    ingredient_id: str
    ingredient_nom: str
    raison: str


class OneTripResponse(BaseModel):
    nb_arrets: int
    distance_totale_km: float
    cout_estime: float
    budget: Optional[float] = None
    ecart: Optional[float] = None
    statut: str
    stops: list[OneTripStop]
    manquants: list[OneTripManquant] = Field(default_factory=list)
    message: str

File: frontend/src/api/marketPanier.ts
import { api } from "./http";

export type PanierItem = {
  ingredient_nom: string;
  quantite: number;
  unite?: string;
  ingredient_id?: string;
};

export type PanierSwapSuggestion = {
  ingredient_nom: string;
  alternative: string;
  economie_estimee: number;
  raison: string;
};

export type PanierCheckResult = {
  cout_estime: number;
  budget: number;
  ecart: number;
  statut: "sous_budget" | "au_budget" | "over_budget";
  swaps: PanierSwapSuggestion[];
};

export function checkPanier(payload: {
  items: PanierItem[];
  budget: number;
  quartier?: string | null;
  lat?: number | null;
  lon?: number | null;
}) {
  return api<PanierCheckResult>("/market/panier-check", {
    method: "POST",
    body: payload,
  });
}

export type OneTripStop = {
  point_de_vente: {
    id: string;
    nom: string;
    type: string;
    latitude: number;
    longitude: number;
  };
  distance_km: number;
  cout_estime: number;
  items: {
    ingredient_id: string;
    ingredient_nom: string;
    quantite: number;
    unite: string;
    prix_unitaire: number;
    cout_estime: number;
  }[];
};

export type OneTripResult = {
  nb_arrets: number;
  distance_totale_km: number;
  cout_estime: number;
  budget?: number | null;
  ecart?: number | null;
  statut: string;
  stops: OneTripStop[];
  manquants: { ingredient_id: string; ingredient_nom: string; raison: string }[];
  message: string;
};

export function planOneTrip(payload: {
  items: PanierItem[];
  lat: number;
  lon: number;
  rayon_km?: number;
  budget?: number;
  profil_id?: string;
}) {
  return api<OneTripResult>("/market/one-trip", {
    method: "POST",
    body: payload,
  });
}
